FLT/AutomorphicForm/QuaternionAlgebra/HeckeOperators/Local.lean

```diff
@@ -0,0 +1,343 @@
+
+-- let F be a (totally real) number field
+variable {F : Type*} [Field F] [NumberField F]
+
+namespace Local
+
+variable {v : HeightOneSpectrum (𝓞 F)}
+
+variable (α : v.adicCompletionIntegers F)
+
+variable (hα : α ≠ 0)
+
+-- Not sure if this could be somewhere else.
+variable (v) {α hα} in
+lemma valuation_lt_one_iff_mem_maximalIdeal {t : adicCompletionIntegers F v} :
+    Valued.v (t : adicCompletion F v) < 1
+    ↔ t ∈ IsLocalRing.maximalIdeal (adicCompletionIntegers F v) :=
+  Iff.trans
+  (Valuation.isEquiv_iff_val_lt_one.mp
+    (Valuation.isEquiv_valuation_valuationSubring Valued.v))
+  (ValuationSubring.valuation_lt_one_iff (adicCompletionIntegers F v) _).symm
+
+variable (v) {α hα} in
+/-- The subgroup `U1 v = GL2.localTameLevel v`. -/
+noncomputable abbrev U1 : Subgroup (GL (Fin 2) (adicCompletion F v)) := (GL2.localTameLevel v)
+
+/- Some lemmas in this section could be placed somewhere else in greater generality. -/
+namespace GL2
+
+/-- The matrix element `diag[α, 1]`. -/
+noncomputable abbrev diag : (GL (Fin 2) (adicCompletion F v)) :=
+  Matrix.GeneralLinearGroup.diagonal (![⟨(α : v.adicCompletion F),
+    (α : v.adicCompletion F)⁻¹, by
+      rw [mul_inv_cancel₀]
+      exact_mod_cast hα, by
+      rw [inv_mul_cancel₀]
+      exact_mod_cast hα⟩, 1])
+
+lemma diag_def :
+    (diag α hα : Matrix (Fin 2) (Fin 2) (adicCompletion F v))
+    = !![↑α, 0; 0, 1] := by
+  rw[diag, Matrix.GeneralLinearGroup.diagonal]
+  ext i j; fin_cases i; all_goals fin_cases j
+  all_goals simp
+
+lemma conjBy_diag {a b c d : adicCompletion F v} :
+    (diag α hα)⁻¹ * !![a, b; c, d] * diag α hα
+    = !![a, (α : v.adicCompletion F)⁻¹ * b; c * α, d] := by
+  simp only [Matrix.coe_units_inv, diag_def, Matrix.inv_def, Matrix.det_fin_two_of, mul_one,
+    mul_zero, sub_zero, Ring.inverse_eq_inv', Matrix.adjugate_fin_two_of, neg_zero, Matrix.smul_of,
+    Matrix.smul_cons, smul_eq_mul, Matrix.smul_empty, Matrix.cons_mul, Nat.succ_eq_add_one,
+    Nat.reduceAdd, Matrix.vecMul_cons, Matrix.head_cons, Matrix.tail_cons, zero_smul,
+    Matrix.empty_vecMul, add_zero, zero_add, Matrix.empty_mul, Equiv.symm_apply_apply,
+    Matrix.add_cons, Matrix.empty_add_empty, EmbeddingLike.apply_eq_iff_eq]
+  rw[inv_mul_cancel₀ ((Subtype.coe_ne_coe).mpr hα), one_mul, one_mul]
+  ring_nf; rw[mul_inv_cancel₀ ((Subtype.coe_ne_coe).mpr hα), one_mul]
+
+/-- The unipotent matrix element `!![1, t; 0, 1]`. -/
+noncomputable def unipotent (t : v.adicCompletion F) : (GL (Fin 2) (adicCompletion F v)) :=
+  letI detInv : Invertible !![1, t; 0, 1].det :=
+  { invOf := 1,
+    invOf_mul_self :=
+      by simp only [Matrix.det_fin_two_of, mul_one, mul_zero, sub_zero],
+    mul_invOf_self :=
+      by simp only [Matrix.det_fin_two_of, mul_one, mul_zero, sub_zero] }
+  Matrix.unitOfDetInvertible !![1, t; 0, 1]
+
+lemma unipotent_def (t : v.adicCompletion F) :
+    (unipotent t : Matrix (Fin 2) (Fin 2) (adicCompletion F v))
+    = !![1, t; 0, 1] := by
+  rw[unipotent]; rfl
```

Comment:
```suggestion
  rw [unipotent]; rfl
```
I'm surprised that we're not linting for bad spaces. Does `simp [unipotent]` solve it?